src/unique.ts

```diff
@@ -0,0 +1,76 @@
+import { Input, hash } from 'fictional'
+import { getHashKey, setHashKey } from './hash'
+
+export type Store<T> = {
+  has(value: T): boolean
+  add(value: T): void
+}
+
+export type UniqueOptions = {
+  /**
+   * The maximum number of attempts to generate a unique value.
+   * If the maximum number of attempts is reached, the method will call attemptReached and return the last generated value.
+   * @default 10
+   */
+  attempts?: number
+  /**
+   * A callback called when the maximum number of attempts is reached.
+   * can be used to log the error in the console of throw an error.
+   *
+   * @example
+   * ```ts
+   * const attemptsReached = (input: Input) => {
+   *  throw new Error(`copycat.unique: maximum number of attempts reached for input ${input}`)
+   * }
+   * copycat.unique(input, { attemptsReached })
+   * ```
+   */
+  attemptsReached?: (input: Input) => void
+}
+
+/**
+ * Will attempt to generate a unique value using the provided method.
+ * It does so by deterministically changing the seed of generation between each attempt.
+ * If the maximum number of attempts is reached, the method will call attemptsReached then return the last generated value.
+ * @example
+ * ```ts
+ * const method = (seed: Input) => {
+ *  return copycat.int(seed, {max: 1000})
+ * }
+ * const store = new Set<ReturnType<typeof method>>()
+ * copycat.unique(input, method,  store, { attempts: 100 })
+ * ```
+ */
+export const unique = <T>(
+  input: Input,
+  method: (input: Input) => T,
+  store: Store<T>,
+  options?: UniqueOptions
+): T => {
+  const { attempts = 10 } = options ?? {}
+  const originalKey = getHashKey()
+  let result: T = method(input)
+
+  try {
+    let attempt = 0
+    let hashInput = hash(input)
+
+    while (store.has(result) && attempt < attempts) {
+      attempt++
+      const newKey = hash(hashInput)
+      // We change the hash key based on the last hash input to make sure that the next result is different.
+      setHashKey(`${newKey}`)
```

Comment:
For record's sake: we discussed on a call investigating whether we need to reset the hash key.

Reply:
For record's sake:

After digging back in we do need to change the `setHashKey` because we don't want to alter the `input` passed to the user `method`.

Basically, I don't change the "input" we send to the method. Because I can't know if the user rely on it or not.
I just use the "hash" to get a pseudo random value that I can use as a hashkey.

Consider this:

copycat.unique(dbValue, (seed) => copycat.int(seed, {min: 0, max: 1000}), store)


In that case the approach of "hashing" the input, and re-hashing it every time to get a change would work.

But what about this case:

copycat.unique(dbValue, (seed) => {
  if (seed.endWith('@mydomain.com') {
     return seed
  }
  return copcycat.email(seed)
}, store)


Then in that case (and many other) altering the "seed" input might break the user function logic.

Another example:
```ts
  const store = new Set<ReturnType<typeof method>>()
  const method = (seed): string => {
    return copycat.scramble(seed)
  }
  const result = copycat.unique(['hello world', 'hello world'], method, store, {
    attempts: 100,
  })
```

In that case to have `scramble` preserving the "shape" of the string, we must keep it as is.